### ISSUE TITLE
Code: 403  Error: Cannot transfer file: Forbidden

### ISSUE THREAD

**Issue by bernd-wechner** [laurent22/joplin#6344]
## Environment

Joplin version: 2.7.15
Platform: Linux
OS specifics: Mint 20.3

## Steps to reproduce

Not so easy. I added a note with webclipper at work, and at home I get errors.

## Describe what you expected to happen
I see this in a banner:

Some items cannot be synchronised. [View them now]

And the items concerned are all images on that page I clipped.

## Logfile

Well from the Dev Console:

```
18:25:27: ResourceFetcher: Could not download resource: 414c15f6b2394391bf046b9cd93ba181 Error: Cannot transfer file: Forbidden
    at newError (node_modules/@joplin/lib/JoplinServerApi.js:192)
    at JoplinServerApi.<anonymous> (node_modules/@joplin/lib/JoplinServerApi.js:207)
    at Generator.next (<anonymous>)
    at fulfilled (node_modules/@joplin/lib/JoplinServerApi.js:5)
    at processTicksAndRejections (internal/process/task_queues.js:95)
```

And in the log file:
```
2022-03-30 18:25:27: "ResourceFetcher: Could not download resource: 414c15f6b2394391bf046b9cd93ba181", "Error: Cannot transfer file: Forbidden
Code: 403
Error: Cannot transfer file: Forbidden
    at newError (/tmp/.mount_JoplinZ3g8TM/resources/app.asar/node_modules/@joplin/lib/JoplinServerApi.js:192:28)
    at JoplinServerApi.<anonymous> (/tmp/.mount_JoplinZ3g8TM/resources/app.asar/node_modules/@joplin/lib/JoplinServerApi.js:207:31)
    at Generator.next (<anonymous>)
    at fulfilled (/tmp/.mount_JoplinZ3g8TM/resources/app.asar/node_modules/@joplin/lib/JoplinServerApi.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
```

I can see the images in my Phone app Joplin (Android) and on my Windows desktop app (portable) at work, but see this error with my desktop app (AppImage, on Linux) at home. Looks like a permission error, but how so? It got the note fine, just the images not.  Joplin server sends a 403.

I should update my Joplin Server actually, it is at v2.6.14 currently and a todo is to update it. Alas am using the dockerless install:

https://github.com/hi-ko/joplin-server-ubuntu-install

and we haven't yet fixed it to handle the latest versions.

**Comment by laurent22** [comment]
What's the error log on the server?

**Comment by bernd-wechner** [comment]
Good question. I have to work out where to find one first ;-).

It's running as a systemd service. `journalctl` reveals little but does show:

```
$ journalctl -u joplin | grep error
Mar 30 15:36:44 nephele run.sh[1730]: 2022-03-30 15:36:44: [error] App: 404: GET /favicon.ico: ::ffff:192.168.0.1: Path not found: favicon.ico
Mar 30 15:36:56 nephele run.sh[1730]: 2022-03-30 15:36:56: [error] App: 404: GET /favicon.ico: ::ffff:192.168.0.1: Path not found: favicon.ico
Mar 30 18:11:53 nephele run.sh[1730]: 2022-03-30 18:11:53: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 46 seconds.
Mar 30 18:11:55 nephele run.sh[1730]: 2022-03-30 18:11:55: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 45 seconds.
Mar 30 18:11:56 nephele run.sh[1730]: 2022-03-30 18:11:56: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 43 seconds.
Mar 30 18:11:57 nephele run.sh[1730]: 2022-03-30 18:11:57: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 42 seconds.
Mar 30 18:11:59 nephele run.sh[1730]: 2022-03-30 18:11:59: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 41 seconds.
Mar 30 18:12:00 nephele run.sh[1730]: 2022-03-30 18:12:00: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 39 seconds.
Mar 30 18:12:04 nephele run.sh[1730]: 2022-03-30 18:12:04: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 35 seconds.
Mar 30 18:12:06 nephele run.sh[1730]: 2022-03-30 18:12:06: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 33 seconds.
Mar 30 18:12:07 nephele run.sh[1730]: 2022-03-30 18:12:07: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 32 seconds.
Mar 30 18:12:14 nephele run.sh[1730]: 2022-03-30 18:12:14: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 25 seconds.
Mar 30 18:12:15 nephele run.sh[1730]: 2022-03-30 18:12:15: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 24 seconds.
Mar 30 18:12:16 nephele run.sh[1730]: 2022-03-30 18:12:16: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 23 seconds.
Mar 30 18:12:17 nephele run.sh[1730]: 2022-03-30 18:12:17: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 22 seconds.
Mar 30 18:12:17 nephele run.sh[1730]: 2022-03-30 18:12:17: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 22 seconds.
Mar 30 18:12:18 nephele run.sh[1730]: 2022-03-30 18:12:18: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 22 seconds.
Mar 30 18:12:19 nephele run.sh[1730]: 2022-03-30 18:12:19: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 20 seconds.
Mar 30 18:12:21 nephele run.sh[1730]: 2022-03-30 18:12:21: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 18 seconds.
Mar 30 18:12:22 nephele run.sh[1730]: 2022-03-30 18:12:22: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 18 seconds.
Mar 30 18:12:26 nephele run.sh[1730]: 2022-03-30 18:12:26: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 13 seconds.
Mar 30 18:12:28 nephele run.sh[1730]: 2022-03-30 18:12:28: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 11 seconds.
Mar 30 18:12:29 nephele run.sh[1730]: 2022-03-30 18:12:29: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 11 seconds.
Mar 30 18:12:36 nephele run.sh[1730]: 2022-03-30 18:12:36: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 3 seconds.
Mar 30 18:12:37 nephele run.sh[1730]: 2022-03-30 18:12:37: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 2 seconds.
Mar 30 18:12:38 nephele run.sh[1730]: 2022-03-30 18:12:38: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 1 seconds.
Mar 30 18:12:39 nephele run.sh[1730]: 2022-03-30 18:12:39: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 1 seconds.
Mar 30 18:12:45 nephele run.sh[1730]: 2022-03-30 18:12:45: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 55 seconds.
Mar 30 18:12:49 nephele run.sh[1730]: 2022-03-30 18:12:49: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 51 seconds.
Mar 30 18:12:56 nephele run.sh[1730]: 2022-03-30 18:12:56: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 44 seconds.
Mar 30 18:13:06 nephele run.sh[1730]: 2022-03-30 18:13:06: [error] App: 429: POST /api/sessions: ::ffff:192.168.0.1: Too many login attempts. Please try again in 34 seconds.
Mar 30 20:05:22 nephele run.sh[1730]: 2022-03-30 20:05:22: [error] App: 404: GET /favicon.ico: ::ffff:192.168.0.1: Path not found: favicon.ico
``` 

(Well it shows MUCH more than that, but that is today's messages (and I clipped it today and synched work, phone and home, 2 of which show the images one of which shows the note, and content but not the images)

It runs as per hi-kos setup, and so lives in `/home/joplin` under which is a `logs` directory, but it's empty.

**Comment by bernd-wechner** [comment]
I have since set up Joplin on a Linux laptop and have the same issue. So Android app is fine, Windows app is fine, Linux app seems to have the problem. The very resources it claims it can't download I can download manually by surfing to my Joplin server and finding them by ID. Logged in with the same credentials. So the server has no problem letting me download them logged in through browser by the Linux app for some reasons it does.

If I didn't know better I'd guess that therefore there is some issue with request mechanism or headers in the Linux app (given it's got the issue).

This is a bother. A side effect it seems is the the Sync never ends, and the Cancel button remains swirling forever and I have a feeling to get a new note to sync up I have to close the Joplin app and start it again and then it goes up before the Cancel button keeps spinning and that warning not remains displayed.

**Comment by bernd-wechner** [comment]
So, I tried the flatpak build and same issue. The list of resources it cannot download because of "Forbidden" is now enormous. Conceivably all of the images I have now. The Exported debug report is clean (does not even mention these failures). 

So I drilled down into debugging and can see the request is made with these two headers:

    "headers": {
        "X-API-AUTH": "ZgnCcFx8ny0u6ikTtrqOG07wE8CnzWWd",
        "X-API-MIN-VERSION": "2.6.0"
    },

which my money says are the culprits. I will drill further.

**Comment by bernd-wechner** [comment]
Lost my money on that one. It's not those headers. I can do this:

```
curl -H "X-API-AUTH: ZgnCcFx8ny0u6ikTtrqOG07wE8CnzWWd" -H "X-API-MIN-VERSION: 2.6.0" URL
```

and the file is downloaded fine. So browser can do it, curl can do it. Joplin can't. Go figure.

So, looking like it's in `shim.fetchBlob` somewhere which can't do it!

**Comment by bernd-wechner** [comment]
And alas drilling into that it's worth instrumenting the code a little and can't do that with the AppImage or FlatPak as they are read only. Be nice to solve this, alas my workstation PC died in sympathy with out COVID infections and quarantine and so I'm waiting on a new motherboard in the hope of resurrecting it. All of which is in part why I have a moment to look at this client side. 

Alas the request is not sniffed on the Network tab of the Dev tools making it hard to catch the finalised complete request. As it has to differ from the curl above and what the browser sends in some way, to be getting a 403 while they don't.

**Comment by bernd-wechner** [comment]
OK, this gets more bizarre with every step. As the Dev tools network tab can't sniff the request I tweaked the proxy server to log them. I tried the curl and checked the request all good. I went to the Joplin client and on this page:

![image](https://user-images.githubusercontent.com/7296506/162122715-116a940e-d5b8-443d-9c47-0958b1126e8d.png)

I click retry on the first Forbidden item and check the requests. And two are logged. The GET that I expect and as it's logged so too it works. I have reproduced it in [Postman](https://web.postman.co) and it works fine. then hot on its tail a POST to `/api/sessions` which I have not yet tested in Postman as I need to work out what is being posted (and my server doesn't log that for me alas). Still the documentation: 

https://joplinapp.org/api/references/rest_api/

is devoid of any mention of **POST/sessions**.

Now that is thus far suspicious as heck. 

1. the Joplin app is sending a POST that isn't documented
2. Being a POST that isn't tested with curl or the browser, it is with pretty certainty the thing that is failing with a 403 error and not the get. That would be consistent with all observations to date.

So more digging to divine what this POST is about and what it's content is. This could of course be entirely unrelated as well, it might be after clicking retry that the APP tries the download and then does some bookkeeping (logging) on the server unrelated to the failure, or possibly even logging it (as I've had no response on where I might find server side logging that too remains a mystery for now.

**Comment by bernd-wechner** [comment]
Loads of errors like this also happening:

```
Too many login attempts. Please try again in 47 seconds. (Code 429)
```

The network tab in dev tools does see a lot of Joplin's traffic just not these resource requests nor the attempted session posts. 

Have been trying back versions of the app ... nothing to report yet.

**Comment by roman-r-m** [comment]
Looks like the POST call to `api/sessions` is coming from here:
https://github.com/laurent22/joplin/blob/dfa22b560e6d6be6dc9283a0eff1b01facc0abd8/packages/lib/JoplinServerApi.ts#L140-L143

`this.sessionId()` does the call.

**Comment by roman-r-m** [comment]
Have you considered running tcpdump on your server? This way you should see all requests and responses without any tweaks.

**Comment by bernd-wechner** [comment]
> tcpdump

Wow, wasn't aware of that utility. I'm sure I've seen it before but didn't spring to mind. I shall experiment. It's on a server with a lot of other traffic so a filter like:

```
sudo tcpdump tcp port 22300
```

helps and I'll see what I can find. I am also trying other back versions of the Linux client and present the terminal (CLI) version. For comparison.

**Comment by roman-r-m** [comment]
In case you didn't know, you can also use wireshark to view recorded dumps. It has nice GUI, filtering capabilities, ect. Very powerful tool.

**Comment by bernd-wechner** [comment]
Wireshark I know. Of course the server has no GUI so tcpdump is a bit better. EIther wait rather low level and I need to read up on interpreting it. One image request for example yields for me:

```
17:14:12.096646 IP _gateway.60890 > joplinserver.22300: Flags [S], seq 3070643095, win 29200, options [mss 1460,sackOK,TS val 3570249362 ecr 0,nop,wscale 7], length 0
17:14:12.096695 IP joplinserver.22300 > _gateway.60890: Flags [S.], seq 2836198304, ack 3070643096, win 65160, options [mss 1460,sackOK,TS val 2967079417 ecr 3570249362,nop,wscale 7], length 0
17:14:12.097108 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 1, win 229, options [nop,nop,TS val 3570249362 ecr 2967079417], length 0
17:14:12.097212 IP _gateway.60890 > joplinserver.22300: Flags [P.], seq 1:749, ack 1, win 229, options [nop,nop,TS val 3570249362 ecr 2967079417], length 748
17:14:12.097232 IP joplinserver.22300 > _gateway.60890: Flags [.], ack 749, win 504, options [nop,nop,TS val 2967079418 ecr 3570249362], length 0
17:14:12.107417 IP joplinserver.22300 > _gateway.60890: Flags [P.], seq 1:7241, ack 749, win 504, options [nop,nop,TS val 2967079428 ecr 3570249362], length 7240
17:14:12.107448 IP joplinserver.22300 > _gateway.60890: Flags [P.], seq 7241:7335, ack 749, win 504, options [nop,nop,TS val 2967079428 ecr 3570249362], length 94
17:14:12.107676 IP joplinserver.22300 > _gateway.60890: Flags [F.], seq 7335, ack 749, win 504, options [nop,nop,TS val 2967079428 ecr 3570249362], length 0
17:14:12.107967 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 1449, win 251, options [nop,nop,TS val 3570249373 ecr 2967079428], length 0
17:14:12.108241 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 2897, win 274, options [nop,nop,TS val 3570249373 ecr 2967079428], length 0
17:14:12.108241 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 4345, win 296, options [nop,nop,TS val 3570249373 ecr 2967079428], length 0
17:14:12.108241 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 5793, win 319, options [nop,nop,TS val 3570249373 ecr 2967079428], length 0
17:14:12.108452 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 7241, win 342, options [nop,nop,TS val 3570249373 ecr 2967079428], length 0
17:14:12.108452 IP _gateway.60890 > joplinserver.22300: Flags [.], ack 7335, win 342, options [nop,nop,TS val 3570249373 ecr 2967079428], length 0
17:14:12.109088 IP _gateway.60890 > joplinserver.22300: Flags [F.], seq 749, ack 7336, win 342, options [nop,nop,TS val 3570249374 ecr 2967079428], length 0
17:14:12.109106 IP joplinserver.22300 > _gateway.60890: Flags [.], ack 750, win 504, options [nop,nop,TS val 2967079430 ecr 3570249374], length 0
```

which is a lot to take in (that's a resource request through the browser). A lot to (re)learn here alas. Up front curious why the odd ports used to send back to the gateway (which is a reverse proxy that takes all HTTPS requests to the joplinserver sending them to port 22300 that the server is listening on.

But to be honest the challenge is from this level reassembling the Request (headers and content) and Response (header and content). One wonders if that is not a wheel invented and with a simpler solution (to sniff the TCP layer and report not TCP packets but HTTP requests and responses - big big time saver if so).

**Comment by bernd-wechner** [comment]
This has proved invaluable. Wheel invented:

https://www.middlewareinventory.com/blog/tcpdump-capture-http-get-post-requests-apache-weblogic-websphere/

Joplin terminal `joplin sync` did not report any errors at all. Alas seems to store data somewhere different to the GUI joplin app, as it can see all my notes but the GUI cannot. The CLI version did end with:

```
Created local items: 1130. Fetched items: 2666/2671.
Downloading resources...
Created local items: 1136. Fetched items: 2671/2671. Completed: 07/04/2022 16:27 (426s)
```

so no complaint at Downloading resources and it did take an age to finish from there. So am search disk for said resources now.

**Comment by bernd-wechner** [comment]
Ha, ha the CLI version of joplin downloaded all the resources without complaint but they are mostly .png files containing just:

```
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 Forbidden</title>
 </head>
 <body>
  <h1>403 Forbidden</h1>
 </body>
</html>
```

hilarious.

**Comment by bernd-wechner** [comment]
Thus far have sniffed what the /api/session POSTs are about, they submit username and password which returns a JSON string containing and id and a user_id which are both seemingly long encoded strings and I am guessing are the session tokens so that a sustained communication can run without the password being transmitted any longer.

Alas on first sync the traces are flooded with responses of the form:

```
{"error":"Too many login attempts. Please try again in 10 seconds."}
```

So I tested the exact post using Postman and that worked. Haven't spotted a GET yet in the trace but have only sniffed a bit. Will let it settle and then see if traffic stops before I try a "retry" on a resource fetch,

These many logins are weird, the app should not be trying to logon so often, once only to establish a session I suspect.

**Comment by roman-r-m** [comment]
> But to be honest the challenge is from this level reassembling the Request (headers and content) and Response (header and content). One wonders if that is not a wheel invented and with a simpler solution (to sniff the TCP layer and report not TCP packets but HTTP requests and responses - big big time saver if so).

This is where Wireshark comes in. Try this - save tcpdump to a file (`tcpdump ... -w joplin.pcap`), transfer it over to your PC/Mac and open in Wireshark. It should automatically ideantify and decode HTTP traffic.

**Comment by bernd-wechner** [comment]
Thanks that's cool too! So far using `tcpdump -A` as per that filter I shared also works, it just has a bit of binary header stuff between the packets. But gives a clear ASCII trace of the HTTP request and response I can look at with `less`

**Comment by roman-r-m** [comment]
> These many logins are weird, the app should not be trying to logon so often, once only to establish a session I suspect.

Yes, from the code it looks like it should save the session but it doesn't for some reason. It also seems like this invalid "too many logon attempts" state isn't being properly cleared.

**Comment by bernd-wechner** [comment]
Yeah and that's a separate issue I should file. Because it now is a roadblock to diagnosing this issue. Havinng cleared by local data and tried the latest Linux app again it now no longer syncs ... it spends all its time grinding on failed login efforts. A few notes come down eventually but nothing like the full set. 

It seems to me, that (sensibly) they server has a rate limiting on login efforts, possibly adaptive (which is when the rate limit is reduced over time as a given client (identified however, but source IP as the simplest) continues hammering the interface. 

And so I have solve this login hammering first. 

As ever, deeply puzzled as I have a Andorid app and Windows client both working fine with the server, which nominally clears all three of them as culpable (not the server entirely though as it may have some issue with the Linux client). On top of which, this issue emerged after a long time of working fine, it just appeared it seems. So I supposed it was a client upgrade that caused it. But as I reproduced it with 2.6.4, 2.5.12 and 2.1.19 Linux AppImage clients and the CLI client has the same issue that seems unlikely.

I shall see if the CLI client has a command that can trigger a resource fetch, Then that would work around this login prompt to see some GET requests on resources in my tcdump captured pcap file (my run last night sees lots of .md GETs but no resource GETs).

**Comment by bernd-wechner** [comment]
OK, this is getting tediously cryptic. With the GUI client out of action until this failure to remember the session tokens issue is resolved (yet to file an issue), I use the CLI client for testing. And here's what I do:

1. Disabled automatic sync on all clients (sync interval disabled) to remove all noise and sync only on request.
2. `joplin sync` on PC to confirm nothing is outstanding to sync (this client has all my notes and syncs well, but as noted all resources are text files of 341 bytes containing a 403 error message in plain text HTML format. 
3. Create new note on my phone with an image in it. This immediately syncs even though I disabled it with sync interval 0. That is already a Joplin bug (TODO file it). It syncs automatically when I create the note. No crisis can work around this (I haven't started tcpdump yet)
4. Now on the joplin server I run `tcpdump -Uw - 'tcp port 22300' | tee sync.pcap | tcpdump -r -`, which starts listening and dumping to sync.pcap and to stdout.
5. On the PC I then run `joplin sync` and it reports `Created local items: 3. Fetched items: 3/3. Completed: 08/04/2022 11:34 (2s)`. Itself a mystery as I have one note, which has one image in it. But we'll get to the breakdown soon.
6. On the joplin server I see the trace on screen and it stops. I stop tcpdump.
7. Back on PC I open `sync.pcap` in wireshark (I have the server mounted with sshfs), and use the `http` filter.
8. Here's what I see:
![image](https://user-images.githubusercontent.com/7296506/162347642-99cc0da0-5491-4825-8bec-d712c94af097.png)

In summary this trace is pretty clear and shows this interaction:

1. joplin opens a sessstion (sends username/password and gets id and user_id which I presume are the session tokens it uses from then on)
2. joplin gets /api/items/root:/info.json:/content, which returns sommething odd a small JSON structure that contains no apprent useful information. This literally:
    ```json
    {
	    "version": 3,
	    "e2ee": {
		    "value": false,
		    "updatedTime": 0
	    },
	    "activeMasterKeyId": {
		    "value": "",
		    "updatedTime": 0
	    },
	    "masterKeys": [],
	    "ppk": {
		    "value": null,
		    "updatedTime": 0
	    }
    }
    ```
3. joplin posts to /api/locks, presumably to get a lock on the sync and avoid clashing sync efforts.
4. joplin now gets a diff (since last sync) from /api/items/root:/:/delta?cursor=NlIk7ucDFv56zu3SfO0NuCbKk0VLiBxK. Which returns:
    ```json
    {
      "items": [
        {
          "id": "cNtEGYZdFvB9YyncE7SsJNpbEKwQMXlG",
          "item_id": "RYXtJngY2JyTlZd98MHKUmldflAe0nuW",
          "item_name": "6111e469245a45e3b3d49ae5087c72c5.md",
          "type": 1,
          "updated_time": 1649381663374,
          "jop_updated_time": 1649381667256
        },
        {
          "id": "wRseRajWDWILMuE0G2xMGA26zCoYo0s8",
          "item_id": "d58CWHPfeXUYU3pb8rjVyEq2D8hGYTEx",
          "item_name": "140b34227d654d29a09b7577a3d3e5e4.md",
          "type": 1,
          "updated_time": 1649381663741,
          "jop_updated_time": 1649381663369
        },
        {
          "id": "3s2zRYUGYf9vkkS4hijS1E5YsEYLeUVe",
          "item_id": "P3mMiCcU0ne6WYrA7jCgeB4xZ01qY9Bp",
          "item_name": ".resource/11400e55936c47e2bb0be43464aa3242",
          "type": 1,
          "updated_time": 1649381673105,
          "jop_updated_time": 0
        },
        {
          "id": "vXPof5NClMZzQy7Qiu2FJFZ839RIRYY1",
          "item_id": "tVxBv0gIIjIObIucFXGQs7vCWiaN6WZ3",
          "item_name": "11400e55936c47e2bb0be43464aa3242.md",
          "type": 1,
          "updated_time": 1649381673202,
          "jop_updated_time": 1649381666696
        }
      ],
      "cursor": "vXPof5NClMZzQy7Qiu2FJFZ839RIRYY1",
      "has_more": false
    }
    ```
5. That's 4 items not 3 (so joplin sync is simplifying things, but again we'll see more on that later)
6. Then there the three GETs all on the .md files but none for the resource - therein lies the hairpulling mystery. Soo the three .md files a fetched. And we see three OK responses as well! So lets look into those.
7. The first OK response containing:
    ```markdown
    id: 140b34227d654d29a09b7577a3d3e5e4
    parent_id: 
    item_type: 1
    item_id: 6111e469245a45e3b3d49ae5087c72c5
    item_updated_time: 1649381656944
    title_diff: "[{\"diffs\":[[1,\"Image test 5\"]],\"start1\":0,\"start2\":0,\"length1\":0,\"length2\":12}]"
    body_diff: "[{\"diffs\":[[1,\"Created in Android\\\n\"]],\"start1\":0,\"start2\":0,\"length1\":0,\"length2\":19}]"
    metadata_diff: {"new":{"master_key_id":"","conflict_original_id":"","share_id":"","is_shared":0,"application_data":"","source_application":"net.cozic.joplin-mobile","source":"joplin","is_todo":0,"markup_language":1,"order":1649381639344,"latitude":"-42.86948340","longitude":"147.30105260","todo_completed":0,"altitude":"0.0000","parent_id":"7b333b17faff4b8992c0a1b99684dcf7","source_url":"","id":"6111e469245a45e3b3d49ae5087c72c5","todo_due":0,"author":""},"deleted":[]}
    encryption_cipher_text: 
    encryption_applied: 0
    updated_time: 2022-04-08T01:34:23.369Z
    created_time: 2022-04-08T01:34:23.369Z
    type_: 13
    ``` 
    which is clearly a diff object of some sort.
8. The second OK response contains the markdown that is the actual note itself with metadata:
    ```markdown
    Image test 5
    
    Created in Android
    
    ![img.jpg.jpeg](:/11400e55936c47e2bb0be43464aa3242)
    
    
    id: 6111e469245a45e3b3d49ae5087c72c5
    parent_id: 7b333b17faff4b8992c0a1b99684dcf7
    created_time: 2022-04-08T01:33:59.354Z
    updated_time: 2022-04-08T01:34:27.256Z
    is_conflict: 0
    latitude: -42.86948340
    longitude: 147.30105260
    altitude: 0.0000
    author: 
    source_url: 
    is_todo: 0
    todo_due: 0
    todo_completed: 0
    source: joplin
    source_application: net.cozic.joplin-mobile
    application_data: 
    order: 1649381639344
    user_created_time: 2022-04-08T01:33:59.354Z
    user_updated_time: 2022-04-08T01:34:27.256Z
    encryption_cipher_text: 
    encryption_applied: 0
    markup_language: 1
    is_shared: 0
    share_id: 
    conflict_original_id: 
    master_key_id: 
    type_: 1
    ```
9. The third OK contains markdown which describes the image resource, albeit with no filename oddly:
    ```markdown
    img.jpg.jpeg
    
    id: 11400e55936c47e2bb0be43464aa3242
    mime: image/jpeg
    filename: 
    created_time: 2022-04-08T01:34:26.696Z
    updated_time: 2022-04-08T01:34:26.696Z
    user_created_time: 2022-04-08T01:34:26.696Z
    user_updated_time: 2022-04-08T01:34:26.696Z
    file_extension: jpeg
    encryption_cipher_text: 
    encryption_applied: 0
    encryption_blob_encrypted: 0
    size: 44166
    is_shared: 0
    share_id: 
    master_key_id: 
    type_: 4
    ```
10. There is NO http request to get the image! Which is a mystery. More on that later.
11. Then it deletes the lock it took out.
12. Finally, and this is a deep mystery, it sends the username and password again and gets another id and user_id. What? Is this another Joplin bug? That the client is requesting a session again rather than as one might expect asking to close the session (expire any session keys).

So, now where did the attempt to download the image go to?

The first observation is that looking in: `~/.config/joplin/resources` we see that image file waas given an id of `11400e55936c47e2bb0be43464aa3242` in the third OK above and in this directory we see a file was added:

```
$ ls -lrt ~/.config/joplin/resources
...
-rw-rw-r-- 1 erkenek erkenek   341 Apr  8 11:25 8516462e50ab418ca86a629d19794b0e.jpeg
drwxr-xr-x 2 erkenek erkenek 28672 Apr  8 11:35 ./
-rw-rw-r-- 1 erkenek erkenek   341 Apr  8 11:35 11400e55936c47e2bb0be43464aa3242.jpeg
drwxr-xr-x 5 erkenek erkenek  4096 Apr  8 11:35 ../
```

with exactly that name and the .jpeg extension. We also see it referred to by the same id in the second OK the note proper which includes the image. But:

```
$ cat ~/.config/joplin/resources/11400e55936c47e2bb0be43464aa3242.jpeg
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 Forbidden</title>
 </head>
 <body>
  <h1>403 Forbidden</h1>
 </body>
</html>
```
So it did try to download it (and its failure mode is different to that of the GUI client, in fact pretty dumb and arguably another bug. The client should report the error arguably on stderr not record it silently as if it were the image content ...

So where did it happen? By inference the CLI client is either not using an HTTP request for the job (and it's hidden among the TCP packets the http filter hid) or wireshark isn't correctly identifying that HTTP request or tcpdump isn't recording it.

Looking at the .pcap file itself suggests the first reason is not the case (there is no evidence of any file download in it), nor the second, so it may be the third (tcpdump doesn't dump it).

So to test this we watch a simple resource request from the browser. I know from sniffing the GUI client earlier that it attempts to request the resources with:

https://my.joplin.server/api/items/root:/.resource/11400e55936c47e2bb0be43464aa3242:/content

and when I do that with tcpdump running, I see:

1. The image displays in the browser.
2. Loaded in wireshark there is a single GET and a single OK response.
3. The OK response can a JPEG file as content.

The deep puzzle then is why I can't see any effort from the joplin CLI client to download that resource and yet it says it's downloading them and it creates a file with faux content (I imagine blithely unaware it has, having received a 403).

Most likely story is it's not using an HTTP request to request but I can't see any evidence at the TCP level in the pcap file of any other effort. I am dumbfounded at present. And hamstrung with a GUI client that doesn't work at all because this session bug.

**Comment by bernd-wechner** [comment]
As aside some useful tools I used:

https://jsonformatter.org/json-pretty-print
https://curl.se/h2c/

sudo npm install -g json

The irony being I usually keep all such notes in Joplin. Which I would if it weren't currently broken here.

**Comment by bernd-wechner** [comment]
I have finally found the culprit. But have yet to understand the mechanism. On March 25th a Spam Filter was installed on the webserver and if I disable that everything works perfectly again. The mystery I have to solve is what configuration int he spam filter blocks just the Linux Joplin client and not the Android or Windows clients I use. That is worth drilling into still and understanding.
